packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -784,18 +784,20 @@ class TextPainter {
   // get rect calls to the paragraph.
   _CaretMetrics _caretMetrics;
 
-  // Holds the TextPosition and caretPrototype the last caret metrics were
-  // computed with. When new values are passed in, we recompute the caret metrics.
-  // only as necessary.
+  // Holds the TextPosition and caretPrototype and width of the TextPainter that
+  // the last caret metrics were computed with. When new values are passed in,
+  // we recompute the caret metrics only as necessary.
   TextPosition _previousCaretPosition;
   Rect _previousCaretPrototype;
+  double _previousWidth;
```

Comment:
I wonder if the caret metrics cache should just be invalidated whenever the text painter has to relayout and its width cache misses?

Reply:
not too familiar with the text painter code, just asking because `TextPainter.markNeedsLayout` seems to invalidate the caret metrics cache.

Reply:
Something like this 554a544?

I like the idea, it seems to make a lot more sense and reduce the need for more state.  Let me get a double check from @HansMuller that it seems like a valid idea.

Reply:
Sounds reasonable to me.